### PR TITLE
chore(sdk): test config cleanup

### PR DESCRIPTION
**Pull request by tonyyli-wandb**
Description
-----------

- Minor partial cleanup of `test/system_tests/conftest.py`
- Refactors implementation of local backend fixture factories (e.g. `user_factory` -> `make_user`), attempting to reduce nested closures / callbacks
- Separates backend fixtures into their own file (`tests/system_tests/backend_fixtures.py` -- this is treated as a `pytest_plugin` by `conftest.py`)

- [x] I updated CHANGELOG.unreleased.md, or it's not applicable

Testing
-------
N/A

### CODE REVIEW

**Comment by tonyyli-wandb** [comment]
* **#9090** <a href="https://app.graphite.dev/github/pr/wandb/wandb/9090?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/wandb/wandb/9090?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: tests/system_tests/backend_fixtures.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+def use_local_wandb_backend(
+    local_wandb_backend: LocalWandbBackendAddress,
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """Fixture that patches WANDB_BASE_URL to point to the local container.
+
+    Unlike `local_wandb_backend`, this is function-scoped, so cannot be used
+    in session-scoped fixtures.
+    """
+    monkeypatch.setenv("WANDB_BASE_URL", local_wandb_backend.base_url)
```
Comment (tonyyli-wandb) [inline tests/system_tests/backend_fixtures.py:97]: **note:** Code **above** here was copied over from `test/system_tests/conftest.py` and should be entirely (or almost entirely) unchanged

File: tests/system_tests/conftest.py [hunk 2]
```diff
@@ -26,6 +21,9 @@
 )
 from .wandb_backend_spy import WandbBackendProxy, WandbBackendSpy, spy_proxy
 
+#: See https://docs.pytest.org/en/stable/how-to/plugins.html#requiring-loading-plugins-in-a-test-module-or-conftest-file
+pytest_plugins = ("tests.system_tests.backend_fixtures",)
+
```
Comment (tonyyli-wandb) [inline tests/system_tests/conftest.py:26]: **note:** This loads `backend_fixtures.py` as a pytest plugin, following guidance from the pytest [docs](https://docs.pytest.org/en/stable/how-to/plugins.html#requiring-loading-plugins-in-a-test-module-or-conftest-file)

File: tests/system_tests/conftest.py [hunk 3]
```diff
@@ -493,115 +415,26 @@ def wandb_verbose(request):
     return request.config.getoption("--wandb-verbose", default=False)
 
 
-@dataclasses.dataclass
-class UserFixtureCommand:
```
Comment (tonyyli-wandb) [inline tests/system_tests/conftest.py:497]: **note:** Renamed to `UserCmd` for brevity and moved to `backend_fixtures.py`

File: tests/system_tests/conftest.py [hunk 4]
```diff
@@ -493,115 +415,26 @@ def wandb_verbose(request):
...
-        ):
-            yield username
-
-            command = UserFixtureCommand(
-                command="down",
-                username=username,
-            )
-            fixture_fn(command)
-
-    yield _user_factory
```
Comment (tonyyli-wandb) [inline tests/system_tests/conftest.py:548]: **note:** Most of the logic for `user_factory` was flattened/absorbed into `BackendFixtureFactory.make_user()` in `backend_fixtures.py`

File: tests/system_tests/conftest.py [hunk 5]
```diff
@@ -493,115 +415,26 @@ def wandb_verbose(request):
...
-
-
-@pytest.fixture(scope="session")
-def fixture_fn_factory():
-    def _fixture_fn_factory(address: LocalWandbBackendAddress):
-        def fixture_util(cmd: UserFixtureCommand) -> bool:
```
Comment (tonyyli-wandb) [inline tests/system_tests/conftest.py:554]: **note:** Most of this was absorbed into `BackendFixtureFactory` in `backend_fixtures.py`.  This seemed like it could be easier to maintain & introspect if flattened, but open to pushback and/or understanding if I missed anything

**Review by estellazx — APPROVED** [review]
thanks for the refactor!

File: tests/system_tests/backend_fixtures.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    local_wandb_backend: LocalWandbBackendAddress,
+    use_local_wandb_backend,
+) -> Generator[BackendFixtureFactory, None, None]:
+    base_url = local_wandb_backend.fixture_service_url
+    with BackendFixtureFactory(base_url, worker_id=worker_id) as factory:
+        yield factory
+
+
+@pytest.fixture
+def backend_importers_fixture_factory(
```
Comment (kptkin) [inline tests/system_tests/backend_fixtures.py:207]: it was in a separate conftest by design

Comment (kptkin) [reply tests/system_tests/backend_fixtures.py:207]: it is only used for importer tests

File: tests/system_tests/backend_fixtures.py
Comment (kptkin) [inline tests/system_tests/backend_fixtures.py:1]: why did you need to do this change?

Comment (tonyyli-wandb) [reply tests/system_tests/backend_fixtures.py:1]: I can move these back (into `conftest.py`) if preferred: let me know if so.

motivation at the time was:
- `conftest.py` already has a lot of different fixture definitions in it already
- it seemed like it would be easier to follow -- and in the future, adapt -- the backend fixture code if it was reorganized into a separate file/plugin

Comment (kptkin) [reply tests/system_tests/backend_fixtures.py:1]: i just wanted to understand the motivation, tho, it is harder to see the diff now with the things you changed, but mainly was curious, since it seemed already in a good shape

Comment (tonyyli-wandb) [reply tests/system_tests/backend_fixtures.py:1]: Yeah it's kind of frustrating that the diff doesn't show the moved code explicitly 😕 , sorry if that was misleading.

None of the fixture logic should really be changed (if I did it right), but since there was discussion about using GQL requests to the backend vs POST requests to the backend fixture service last week, I'm hoping the implementation is easier to rewrite later if needed

File: tests/system_tests/backend_fixtures.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+    _client: httpx.Client = field(init=False)
+    _cleanup_stack: deque[FixtureCmd] = field(default_factory=deque, init=False)
+
+    def __post_init__(self, service_url: str):
+        self._client = httpx.Client(
+            base_url=service_url,
+            # event_hooks={"response": [httpx.Response.raise_for_status]},
+            # timeout=None,
+        )
```
Comment (kptkin) [inline tests/system_tests/backend_fixtures.py:146]: you have commented out code

File: tests/system_tests/backend_fixtures.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        endpoint = str(self._client.base_url.join(path))
+        # FIXME: Figure out how SDK team preferences/conventions for replacing print statements
+        print(f"Triggering fixture on {endpoint!r}: {data!r}", file=sys.stderr)  # noqa: T201
+        try:
+            response = self._client.post(path, json=data)
+            response.raise_for_status()
+        except httpx.HTTPStatusError as e:
+            # FIXME: Figure out how SDK team preferences/conventions for replacing print statements
+            print(e.response.json(), file=sys.stderr)  # noqa: T201
+
```
Comment (kptkin) [inline tests/system_tests/backend_fixtures.py:186]: fixmes here

Comment (tonyyli-wandb) [reply tests/system_tests/backend_fixtures.py:186]: Right thanks for the reminder -- 🙏

`print` statement was kept from existing code, but are there SDK team preferences for what to replace it with?  (Eg if there are logging/reporting patterns used elsewhere)
